### ISSUE TITLE
find_package(ZXing) does not work if ZXing installed in a custom location

### ISSUE THREAD

**Issue by tanius** [nu-book/zxing-cpp#132]
I made a minimal example application that includes a ZXing header file:

```
# CMakeListe.txt
cmake_minimum_required(VERSION 3.5)
project(minimal-zxing-demo LANGUAGES CXX)
find_package(ZXing REQUIRED)
add_executable(minimal-zxing-demo main.cpp)

target_link_libraries(minimal-zxing-demo ZXing)
# ▲ Forgot to mention this line in the original report; edited in now.
```
```
// main.cpp
#include <iostream>
#include <ZXing/BarcodeFormat.h>

using namespace std;
int main() { cout << "Hello World!" << endl; return 0; }
```

This builds successfully if ZXing has been installed system-wide, as follows:

```
cd /path/to/zxing-cpp/sources && mkdir build && cd build
cmake ..
make
sudo make install

cd /path/to/minimal-zxing-demo/sources && mkdir build && cd build
cmake ..
make
```

However, this fails to build if ZXing has instead been installed with a custom install prefix, as follows. If I start from fresh and execute these commands:

```
cd /path/to/zxing-cpp/sources && mkdir build && cd build
cmake .. -DCMAKE_INSTALL_PREFIX=/path/to/zxing-cpp/sources/export
make
make install

cd /path/to/minimal-zxing-demo/sources && mkdir build && cd build
cmake .. -DCMAKE_PREFIX_PATH=/path/to/zxing-cpp/sources/export/lib/cmake/ZXing
make
```

then, for minimal-zxing-demo, the `cmake` step succeeds (means CMake finds the ZXing package) but the following `make` step fails with `ZXing/BarcodeFormat.h: No such file or directory`. The same behaviour is seen when setting `ZXing_DIR` instead of CMAKE_PREFIX_PATH.

Full build output with the error messages is as follows:

```
$ cmake .. -DCMAKE_PREFIX_PATH=[...]/zxing-cpp/export/lib/cmake/ZXing
-- The CXX compiler identification is GNU 9.2.1
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: [...]/minimal-zxing-demo/build

$ make
Scanning dependencies of target minimal-zxing-demo
[ 50%] Building CXX object CMakeFiles/minimal-zxing-demo.dir/main.cpp.o
[...]/main.cpp:2:10: fatal error: ZXing/BarcodeFormat.h: No such file or directory
    2 | #include <ZXing/BarcodeFormat.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/minimal-zxing-demo.dir/build.make:63: CMakeFiles/minimal-zxing-demo.dir/main.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:73: CMakeFiles/minimal-zxing-demo.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

In a different project that actually calls ZXing code, I added `include_directories(/path/to/zxing-cpp/sources/export)` to the `CMakeLists.txt` file to get past the error above. Then compilation succeeds but afterwards the linker cannot find `-lZXing`. So it appears that the CMake `find_package(…)` mechanism does not work to find any ZXing files if ZXing has been installed into a custom location using `CMAKE_INSTALL_PREFIX`.

But this use of custom install locations works for me in another project and I confirmed the difference in find_package(ZXing) behavior with the minimal demo above and. So I hope I made sufficiently sure this is not just due to my limited knowledge of CMake (but I don't exclude that either …).

## Environment

* Ubuntu Linux 19.10
* cmake 3.13.4
* latest ZXing source code (commit ed55911)

**Comment by axxel** [comment]
Your `CMakeLists.txt` is missing a line:
```
target_link_libraries(minimal-zxing-demo ZXing::ZXing)
```

**Comment by tanius** [comment]
Oh, of course! That solves it.

Actually I had this line in my `CMakeLists.txt` but forgot it in the example above (now edited in):

```
target_link_libraries(minimal-zxing-demo ZXing)
```

And that was causing my issue. Building my example with ZXing installed in a custom location works when replacing that line with this one, as axxel proposed:

```
target_link_libraries(minimal-zxing-demo ZXing::ZXing)
```

## Explanations

If anyone runs into the same problems, here's an explanation of the CMake mechanisms at play.

`ZXing::ZXing` is the proper target name of the ZXing-Cpp library, defined [here](https://github.com/nu-book/zxing-cpp/blob/fdc0941CMakeLists.txt#L74) for the CName namespace component `ZXing::` and [here](https://github.com/nu-book/zxing-cpp/blob/fdc0941f7c86dfffc32e2912dc1b17442f871d98/core/CMakeLists.txt#L433) for the exported target name `ZXing`. The plain exported target name `ZXing` could be used if you build ZXing as a sub-project with the same CMake call, but the installed version is only under target name `ZXing::ZXing` because the namespace component is added when installing. And since `find_package(…)` is used *only* with installed packages, that is the relevant name here.

So, first learning, always look up the exact [IMPORTED target name](https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#imported-targets) provided by a library and call `target_link_libraries(…)` in your code with _that_. It avoids issues.

The issues can come from a fallback behavior of CMake. In my original call of `target_link_libraries(minimal-zxing-demo ZXing)`, CMake will not find an IMPORTED target name `ZXing` and then it does this:

> if a target was not found with a matching name, the name […] [is] considered to refer to a file on disk. [Except that there will be] a FATAL_ERROR if a link dependency contains double-colons but is not an IMPORTED target or an ALIAS target.
> ([source](https://cmake.org/cmake/help/v3.0/policy/CMP0028.html); note that the described policy 0028 is usually set and in CMake 3.0.2 there will be a warning otherwise)

Since `ZXing` does not contain a double-colon `::`, it is "considered to refer to a file on disk`. This is equivalent to a linker argument `-lZXing`, refering to a library with filename `libZXing.so` under Linux for example.

Now if ZXing has been installed system-wide, the linker will find that file in the usual library search paths. Similarly, the system will know the include paths of the library of that file name and will hand them to the compiler. However, if ZXing had been installed into a custom location, the compiler and linker does not know where to find a library `libZXing.so` because nobody told it. The fact that I had also called `find_package(ZXing)` does not tell the compiler anything about the paths to find this `libZXing.so` on disk because it only talks about target `ZXing::ZXing`, not `ZXing`.

Of course it would be possible to also export the non-namespaced target `ZXing`. In that case, `find_package(ZXing)` would help the compiler and linker find `target_link_libraries(minimal-zxing-demo ZXing)` even in a custom location. But that approach would not make anything better because namespaced target names are the established practice for using CMake and have the benefit of making `target_link_libraries(…)` fail if such a name is not provided by a package ([see](https://cmake.org/cmake/help/latest/policy/CMP0028.html)). Which is exactly the behavior we want to see here: no fallback of `target_link_libraries(…)` to library filenames. So, nothing wrong in ZXing-Cpp CMake.

**tl;dr** For hassle-free "modern" CMake usage, call `target_link_libraries(…)` exclusively with double-colon library target names. For ZXing-Cpp, do it like this: `target_link_libraries(minimal-zxing-demo ZXing::ZXing)`

I'll send a PR for documenting the proper use of `target_link_libraries(…)` in the README.

**Comment by axxel** [comment]
Glad you could solve your issue. And thanks for the offer to contribute to the README but I consider this how-to-cmake background info to be 'general knowledge' and would rather not include that in the README, which is already too wordy for my taste. Also there is already the `example` directory showing a working `CMakeListst.txt` to have look at.

**Comment by tanius** [comment]
Well, thank you for your assistance with this issue :slightly_smiling_face: 

I prefer wordy documentation about using tools that are complex and hard to debug like CMake. It would not tell me something like "Warning: no imported target from package ZXing was used. Available targets: ZXing::ZXing." I didn't find a command to list targets and files imported from a package either. But not sure I want to even propose that stuff to the CMake folks.

But sure, documentation styles differ. I guess anyone with the same problem will find this issue via search engine now, so it's fine.
